pkg/cloudprovider/controller/nodecontroller.go

```diff
@@ -421,30 +373,108 @@ func (s *NodeController) checkNodeReady(node *api.Node) *api.NodeCondition {
+			addr := net.ParseIP(node.Name)
+			if addr != nil {
+				address := api.NodeAddress{Type: api.NodeLegacyHostIP, Address: addr.String()}
+				node.Status.Addresses = []api.NodeAddress{address}
+			} else {
+				addrs, err := s.lookupIP(node.Name)
+				if err != nil {
+					glog.Errorf("Can't get ip address of node %s: %v", node.Name, err)
+				} else if len(addrs) == 0 {
+					glog.Errorf("No ip address for node %v", node.Name)
+				} else {
+					address := api.NodeAddress{Type: api.NodeLegacyHostIP, Address: addrs[0].String()}
+					node.Status.Addresses = []api.NodeAddress{address}
+				}
+			}
+		}
+	}
+	return nodes, nil
+}
+
+// MonitorNodeStatus verifies node status are constantly updated by kubelet, and if
+// not, post node NotReady status. It also evicts all pods if node is not ready for
+// a long period of time.
+func (s *NodeController) MonitorNodeStatus() error {
+	nodes, err := s.kubeClient.Nodes().List()
 	if err != nil {
 		return err
 	}
-	for _, pod := range pods.Items {
-		if pod.Status.Host != nodeID {
-			continue
+	for i := range nodes.Items {
+		node := &nodes.Items[i]
+		// Precompute all condition times to avoid deep copy of node status (We'll modify node for
+		// updating, and NodeStatus.Conditions is an array, which makes assignment copy not useful).
+		latestConditionTime := s.latestConditionTime(node, api.NodeReady)
+		lastProbeTime := s.getCondition(node, api.NodeReady).LastProbeTime
+		lastTransitionTime := s.getCondition(node, api.NodeReady).LastTransitionTime
+		var gracePeriod time.Duration
+		if latestConditionTime == node.CreationTimestamp {
+			gracePeriod = nodeStartupGracePeriod
+		} else {
+			gracePeriod = nodeMonitorGracePeriod
 		}
-		glog.V(2).Infof("Delete pod %v", pod.Name)
-		if err := s.kubeClient.Pods(pod.Namespace).Delete(pod.Name); err != nil {
-			glog.Errorf("Error deleting pod %v: %v", pod.Name, err)
+		// NodeReady condition was last set longer ago than gracePeriod, so update it to ConditionNone
+		// (regardless of its current value) in the master, without contacting kubelet.
+		if util.Now().After(latestConditionTime.Add(gracePeriod)) {
+			readyCondition := s.getCondition(node, api.NodeReady)
+			if readyCondition == nil {
+				glog.V(2).Infof("node %v is never updated by kubelet. latestConditionTime is %v",
+					node.Name, latestConditionTime)
+				node.Status.Conditions = append(node.Status.Conditions, api.NodeCondition{
+					Type:               api.NodeReady,
+					Status:             api.ConditionNone,
+					Reason:             fmt.Sprintf("Kubelet never posted node status"),
+					LastProbeTime:      util.Now(),
+					LastTransitionTime: util.Now(),
+				})
+			} else {
+				// Note here the out-dated condition can be the one posted by nodecontroller
+				// itself before. We keep posting the status to keep LastProbeTime fresh.
+				glog.V(2).Infof("node %v hasn't been updated for %v: out-dated condition is %+v",
+					node.Name, gracePeriod, readyCondition)
```

Comment:
instead of gracePeriod I think it would be better to use now-latestConditionTime
also, "outdated readyCondition is" rather than "outdated condition is"